privacyidea/lib/container.py

```diff
@@ -0,0 +1,782 @@
+
+    :param container_serial: The serial of the container
+    :param token_serial: The serial of the token
+    :param user: The user adding the token
+    :param user_role: The role of the user ('admin' or 'user')
+    :return: True on success
+    """
+    container = find_container_by_serial(container_serial)
+    # Check if user is admin or owner of container
+    _check_user_access_on_container(container, user, user_role)
+
+    # Get the token object
+    token = get_tokens_from_serial_or_user(token_serial, None)[0]
+
+    # Check if the token is in a container
+    old_container = find_container_for_token(token_serial)
+
+    # Check if admin/user is allowed to add the token to the container
+    if user_role == "admin" or token.user == user:
+        if old_container:
+            # Remove token from old container (raises PolicyError if user is not allowed to edit the old container)
+            remove_token_from_container(old_container.serial, token_serial, user, user_role)
+            log.info(f"Adding token {token.get_serial()} to container {container_serial}: "
+                     f"Token removed from previous container {old_container.serial}.")
+        res = container.add_token(token)
+
+    else:
+        raise PolicyError(f"User {user} is not allowed to add token {token.get_serial()} "
+                          f"to container {container_serial}.")
+    return res
+
+
+def add_multiple_tokens_to_container(container_serial, token_serials, user: User = None, user_role="user",
+                                     allowed_realms=[]):
+    """
+    Add the given tokens to the container with the given serial. Raises a ResourceNotFoundError if the container does
+    not exist. If a token is already in a container it is removed from the old container.
+    A user is only allowed to add a token to a container if the user is an admin or the owner of both. If the token is
+    already in a container, the user also has to be the owner of the old container.
+
+    :param container_serial: The serial of the container
+    :param token_serials: A list of token serials to add
+    :param user: The user adding the tokens
+    :param user_role: The role of the user ('admin' or 'user')
+    :param allowed_realms: A list of realms the admin is allowed to add tokens to, optional
+    :return: A dictionary in the format {token_serial: success}
+    """
+    # Raises ResourceNotFound if container does not exist
+    find_container_by_serial(container_serial)
+
+    ret = {}
+    for token_serial in token_serials:
+        # Check if admin is allowed to add the token to the container
+        if user_role == "admin" and allowed_realms:
+            token_realms = get_realms_of_token(token_serial)
+            matching_realms = list(set(token_realms).intersection(allowed_realms))
+            if len(matching_realms) == 0:
+                ret[token_serial] = False
+                log.error(
```

Comment:
as mentioned above, please use `info` or `warning` here since this is (kind of) expected behaviour. `error` should be reserved for errors which impact the functionality of the server.